vfs: fix unicode normalization on macOS - fixes #7072

nielash opened the pull request:
#### What is the purpose of this change?

Before this change, the VFS layer did not properly handle unicode normalization, which caused problems particularly for users of macOS. While attempts were made to handle it with various `-o modules=iconv` combinations, this was an imperfect solution, as no one combination allowed both NFC and NFD content to simultaneously be both visible and editable via Finder.

After this change, the VFS supports `--no-unicode-normalization` (default `false`) via the existing `--vfs-case-insensitive` logic, which is extended to apply to both case insensitivity and unicode normalization form.

This change also adds an additional flag, `--vfs-block-norm-dupes`, to address a probably rare but potentially possible scenario where a directory contains multiple duplicate filenames after applying case and unicode normalization settings. In such a scenario, this flag (disabled by default) hides the duplicates. This comes with a performance tradeoff, as rclone will have to scan the entire directory for duplicates when listing a directory. For this reason, it is recommended to leave this disabled if not needed. However, macOS users may wish to consider using it, as otherwise, if a remote directory contains both NFC and NFD versions of the same filename, an odd situation will occur: both versions of the file will be visible in the mount, and both will appear to be editable, however, editing either version will actually result in only the NFD version getting edited under the hood. `--vfs-block-norm-dupes` prevents this confusion by detecting this scenario, hiding the duplicates, and logging an error, similar to how this is handled in `rclone sync`.

#### Was the change discussed in an issue or in the forum before?

- #7072 
- https://forum.rclone.org/t/possible-special-character-encoding-issue-on-macos/39048/30
- https://forum.rclone.org/t/some-unicode-forms-break-mount-on-macos-with-fuse-t/36403

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

kapitainsky commented:
Thank you very much for tackling this pesky issue with your PR! macOS users will be grateful:)

I have been testing this feature for the last few days and it works like a charm with whatever NFC/NFD objects I throw at mount or remote. And everything is accessible on macOS using both terminal and Finder.

This looks like true solution for years old problem with NFC/NFD on macOS.  It works perfectly with both macFUSE and FUSE-T. No more any snake oil `iconv` options needed - it never really fully worked anyway.

@ncw I see that #7072 is on `v1.66` [milestones list](https://github.com/rclone/rclone/milestone/54) (it was also on `v1.65`:)) so hopefully you can include it in the upcoming rclone release.

kapitainsky commented:
Is there any issue stopping this change from being merged?

kapitainsky commented:
I am sorry for pestering but as this is something causing issues in my daily rclone usage I keep a close eye on this possible fix.

@ncw Has this change have a chance to be included in the next rclone release? Or is there still something needed required to do it?

ncw commented on vfs/dir.go:
```diff
@@ -514,6 +515,41 @@ func (d *Dir) _readDir() error {
+		normalizedNames := make(map[string]struct{}, entries.Len())
+		filteredEntries := make(fs.DirEntries, 0)
+		for _, e := range entries {
+			normName := fmt.Sprintf("%s-%T", operations.ApplyTransforms(ctx, e.Remote()), e) // include type to track objects and dirs separately
```
Maybe we should have a version of ApplyTransforms we say explicitly which transforms we want - might save mucking about with contexts

nielash replied:
Interesting idea! I'm happy to make one. I didn't do it yet as I'm not sure how much it would buy us here, since we would still need to at least read the global config to get the `ci.IgnoreCaseSync` and `ci.NoUnicodeNormalization` values (and we'd still need some way of setting those values from the tests).

ncw replied:
I think it is a bit of a code smell to to having to adjust a `ctx` - we should use function parameters for passing values to functions.

From the [Go docs](https://pkg.go.dev/context) this is what contexts are for:

> Package context defines the Context type, which carries deadlines, cancellation signals, and other request-scoped values across API boundaries and between processes.

We've wedged configuration in there but I think changing it instead of making a function with more parameters is probably against the spirit of what context is for!

nielash replied:
Got it -- made this change.

ncw reviewed (commented):
I think this looks sound :-)

I have some quibbles with the use of the global config though - see inline!

Thank you !

nielash commented:
@ncw very helpful feedback, thank you! I think `fs.AddConfig` will work well for the actual code, but I'm not sure it will work for passing options from the tests, because it seems that the VFS methods do not actually take a `ctx` function parameter -- they all just use `context.TODO()`. Should I try to change them to take a function param?

Another option is maybe to make a new VFS option, although it seems not ideal to have lots of options that kind of do the same thing (like we kind of already have with `vfs.Opt.CaseInsensitive` and `ci.IgnoreCaseSync`). 

Whatever you think is best is fine with me 🙂

nielash commented:
Here's a revised version that addresses most of the feedback. The production code now uses `fs.AddConfig`. The one change I did not fully make is changing the _test_ code to use `fs.AddConfig`, because of what I mentioned in my [previous comment](https://github.com/rclone/rclone/pull/7620#issuecomment-1971386456). For now I have addressed that concern in a different way (by restoring the prior value after the test), but I remain open to different solutions if you prefer!

nielash commented on vfs/dir.go:
```diff
@@ -767,15 +805,22 @@ func (d *Dir) stat(leaf string) (Node, error) {
-	if !ok && d.vfs.Opt.CaseInsensitive {
-		leafLower := strings.ToLower(leaf)
+	ctx := context.TODO()
+	ci := fs.GetConfig(ctx)
```
Made these changes.

nielash commented on cmd/cmount/mount.go:
```diff
@@ -116,13 +116,14 @@ func mountOptions(VFS *vfs.VFS, device string, mountpoint string, opt *mountlib.
 		if !findOption("modules=iconv", options) {
 			iconv := "modules=iconv,from_code=UTF-8,to_code=UTF-8-MAC"
 			options = append(options, "-o", iconv)
 			fs.Debugf(nil, "Adding \"-o %s\" for macOS", iconv)
```
Done :)

nielash commented on vfs/dir.go:
```diff
@@ -514,6 +515,41 @@ func (d *Dir) _readDir() error {
+		ciLock.Lock()
+		defer ciLock.Unlock()
+		ctx := context.TODO()
+		ci := fs.GetConfig(ctx)
```
Made this change.

nielash commented on vfs/vfs_case_test.go:
```diff
@@ -86,6 +88,8 @@ func TestCaseSensitivity(t *testing.T) {
 	assert.NotEqual(t, err, ENOENT)
 
 	// Run the same set of checks with case-Sensitive VFS, for comparison.
+	ci := fs.GetConfig(ctx)
```
I just removed this as it's no longer needed.

nielash commented on vfs/vfs_case_test.go:
```diff
@@ -160,3 +164,34 @@ func assertFileAbsentVFS(t *testing.T, vfs *VFS, name string) {
+	assertFileDataVFS(t, vfs, nfd, "data2")
+
+	// change ci.NoUnicodeNormalization to true and verify that only file1 is found
+	ci := fs.GetConfig(ctx)
```
I addressed this in a different way for now -- see [comment here](https://github.com/rclone/rclone/pull/7620#issuecomment-1975656951) about why, and some possible alternatives.

ncw reviewed (commented):
I think your use of context in the tests is fine! The VFS doesn't have contexts and that is because the underlying Go interfaces don't take contexts which is unfortunate.

However on further reflection I think we should try to avoid having to change the context in the main code - see the notes I put inline! Reading it is fine and expected but changing it is not ideal.

nielash commented:
Makes sense! Here's a new version with this changed.

nielash commented on vfs/dir.go:
```diff
@@ -767,15 +796,18 @@ func (d *Dir) stat(leaf string) (Node, error) {
 
-	if !ok && d.vfs.Opt.CaseInsensitive {
-		leafLower := strings.ToLower(leaf)
+	_, ci := fs.AddConfig(context.TODO())
```
Should this be changed back to `fs.GetConfig` now? (Does it matter?)

ncw replied:
Yes it should be just `fs.GetConfig` if we don't need to change the config

ncw reviewed (approved):
This looks great now - thank you :-)

I'll merge it.
